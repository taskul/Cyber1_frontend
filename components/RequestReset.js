import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import useForm from '../lib/useForm';
import Form from './styles/Form';
import { CURRENT_USER_QUERY } from './User';
import Error from './ErrorMessage';

// sendUserPasswordResetLink is part of Keystone API
const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    sendUserPasswordResetLink(email: $email) {
      #   Data we are requesting to get back
      code
      message
    }
  }
`;

export default function RequestReset() {
  const { inputs, handleChange, resetForm } = useForm({
    email: '',
  });
  const [signup, { data, error, loading }] = useMutation(
    REQUEST_RESET_MUTATION,
    {
      variables: inputs,
      // refetch the currently logged in user
      // refetchQueries: [{ query: CURRENT_USER_QUERY }],
    }
  );
  async function handleSubmit(e) {
    e.preventDefault(); // stop the form from submitting
    // adding catch at the end will prevent browser console from displaying error messsages that related to this, because we are rendering below errors with useMutation, but then signup() also returns the same errors, so we also need to catch those.
    const res = await signup().catch(console.error);
    resetForm();
    // Send the email and password to the graphqlAPI
  }

  return (
    // specifying the method as post prevent our from from posting user password into request string queries
    <Form method="POST" onSubmit={handleSubmit}>
      <h2>Reset Password</h2>
      <Error error={error} />
      <fieldset>
        {/* sendUserPasswordResetLink returns null when mutation is used with Keystone */}
        {data?.sendUserPasswordResetLink === null && (
          <p>Reset password email was sent!</p>
        )}

        <label htmlFor="email">
          Email
          <input
            type="email"
            name="email"
            placeholder="Email"
            autoComplete="email"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>
        <button type="submit">Send a Reset Password Request</button>
      </fieldset>
    </Form>
  );
}

export { REQUEST_RESET_MUTATION };
