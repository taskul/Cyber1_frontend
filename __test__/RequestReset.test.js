import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import RequestReset, {
  REQUEST_RESET_MUTATION,
} from '../components/RequestReset';

const email = 'cat@cats.com';
const mocks = [
  {
    request: {
      query: REQUEST_RESET_MUTATION,
      variables: { email },
    },
    result: {
      // the way that Keystone works is that if the requst is successful we get null back
      data: { sendUserPasswordResetLink: null },
    },
  },
];

describe('<RequestReset/>', () => {
  it('renders and matches snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <RequestReset />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('calls the mutation when submitte', async () => {
    const { container } = render(
      <MockedProvider mocks={mocks}>
        <RequestReset />
      </MockedProvider>
    );
    // type into the boxes
    // simulating user using the website
    // type means typing into the field, and we'll use "placeholder" as a way to find the field because that is what user would use to spot the field. Field label also would work
    // we use Regex for email field /email/i means includes word email and is case insensitive
    userEvent.type(screen.getByPlaceholderText(/email/i), email);
    // click submit
    userEvent.click(screen.getByText(/Send a Reset Password Request/));
    const success = await screen.findByText(/Reset password email was sent!/i);
    screen.debug(success);
  });
});
