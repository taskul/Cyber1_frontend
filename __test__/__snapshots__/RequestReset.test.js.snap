// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset/> renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 jndoER"
    method="POST"
  >
    <h2>
      Reset Password
    </h2>
    <fieldset>
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          name="email"
          placeholder="Email"
          type="email"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Send a Reset Password Request
      </button>
    </fieldset>
  </form>
</div>
`;
